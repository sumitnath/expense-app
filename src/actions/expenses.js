import database from '../firebase/firebase';
import uuid from 'uuid';


export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = {
            description: description,
            note: note,
            amount: amount,
            createdAt: createdAt
        };
        return database.ref('expenses').push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};

//Remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id: id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    }
}

// Edit Expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            })
    }
}

// SET_Expense
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});
// export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach(childSnapshot => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            });
    }
}