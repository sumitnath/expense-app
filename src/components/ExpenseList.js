import React from "react";
import { connect } from 'react-redux'
import ExpenseListItem from "./ExpenseListItem";
import SelectExpenses from '../selector/expenses'

export const ExpenseList = (props) => (
<div>
<h1>Expense List</h1>
{
    props.expenses.length === 0 ? (<p>No expenses</p>) :  (props.expenses.map((expense)=>{
        return  <ExpenseListItem key ={expense.id}{...expense}/>
     }))
}

</div>
)
const mapStateToProps = (state)=>{
    return {
        expenses: SelectExpenses(state.expenses,state.filters)
    }
}
export default  connect (mapStateToProps)(ExpenseList)
