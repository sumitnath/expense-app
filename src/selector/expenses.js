import moment from "moment"
// get visible data
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createAtMoment = moment(expense.createdAt)
            // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate
        const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate
        const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}