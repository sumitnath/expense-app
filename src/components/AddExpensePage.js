import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses'
import { startAddExpense } from '../actions/expenses';
// getting action geberator 
export class AddExpensePage extends React.Component{
  onSubmit=(expense)=>{
    //console.log(expense);
   // props.dispatch(addExpense(expense))
   this.props.startAddExpense(expense)
    this.props.history.push('/')
  };
  render(){
  return (
    <div>
    <h1> Add Expenses</h1>
    <ExpenseForm 
    onSubmit={this.onSubmit} />
  </div>
  )
 } 
}

  const mapDispatchToProps = (dispatch) =>{
    return {
      startAddExpense : (expense)=> dispatch(startAddExpense(expense))
        }
  }
//to pass the data to redux store we have to import connect 
// we donot need anything from the state we will keep 1st component of connect empty

export default connect(undefined, mapDispatchToProps)(AddExpensePage)