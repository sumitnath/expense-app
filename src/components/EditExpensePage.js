import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense} from '../actions/expenses'
import {removeExpense} from '../actions/expenses'

export class EditExpensePage extends React.Component {
  
 editExpense = (expense)=>{
      console.log('update',expense);
   //  this.props.dispatch(editExpense(this.props.expense.id,expense))
   this.props.increment(this.props.expense.id,expense)
     this.props.history.push('/')   
    }
  removeExpense =()=>{ 
    this.props.decrement({id:this.props.expense.id} );
    this.props.history.push('/')
  }
  render(){
return (
  <div>
  <h3> Edit Your expense</h3> 
    
    <ExpenseForm 
    expense ={this.props.expense}
    onSubmit={this.editExpense}
    />
    
  <button onClick ={this.removeExpense}>Remove</button>
    
  </div>
   )
  }
}
const mapDispatchToProps = (dispatch,props) => {
    return {
        increment: (id,expense) => dispatch(editExpense(id, expense)),
        decrement: (data) => dispatch(removeExpense(data))   
    }
  }

const mapStateToProps =(state,props ) =>{
  return{
  expense: state.expenses.find((expense)=>
   {return expense.id === props.match.params.id})
   }
  }
    
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)


