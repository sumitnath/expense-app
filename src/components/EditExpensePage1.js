import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses'
import { removeExpense } from '../actions/expenses'

// changing stateless function component to class based component in EditExpensePage 
const EditExpensePage = (props) => {
    console.log(props)
    return ( <div>
        <h3> Edit Your expense </h3>  
        <p> The id is: { props.match.params.id } </p>
         <ExpenseForm expense = { props.expense }
        onSubmit = {
            (expense) => {
                console.log('update', expense);
                props.dispatch(editExpense(this.props.expense.id,expense))
                props.history.push('/')
            }
        }
        />

        <button onClick = {() => {
                props.dispatch(removeExpense({ id: props.expense.id }));
             props.history.push('/')
            }
        } > Remove </button>

         </div>
    );
}


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => { return expense.id === props.match.params.id })
    }
}



export default connect(mapStateToProps)(EditExpensePage)