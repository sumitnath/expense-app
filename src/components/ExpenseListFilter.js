import React from "react";
import { connect } from "react-redux";
import{setTextFilter,sortByDate,sortByAmount, setStartDate,setEndDate} from '../actions/filters'
import { DateRangePicker } from "react-dates";

// converting the stateless component to class component
//const ExpenseListFilter = (props) => (
export class ExpenseListFilter extends React.Component {
   state ={
      calenderFocused: null
   }
   onDatesChange =({startDate:startDate,endDate:endDate})=>{
     this.props.setStartDate(startDate);
     this.props.setEndDate(endDate);
   }
   onFocusChange = (calenderFocused)=>{
     this.setState(()=>({calenderFocused:calenderFocused}))
   }
   onTextChange = (e)=>{
             this.props.setTextFilter(e.target.value)
          //console.log(e.target.value)
         }
   onSortChange =   (e)=>{
          if(e.target.value === 'date'){
             this.props.sortByDate()
          }else if(e.target.value === 'amount'){
             this.props.sortByAmount()
          }
         }    
   render(){
      return(
         <div>
         <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
         <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
         <option value ="date">Date</option>
         <option value="amount">Amount</option>
         </select>
         <DateRangePicker
           startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          // startDateId={this.props.filters.startDate.toString()} // PropTypes.string.isRequired,
           endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          // endDateId={this.props.filters.endDate.toString()}// PropTypes.string.isRequired,
           onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
           focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
           onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
           numberOfMonths= {1}
           isOutsideRange= {()=>false}
           showClearDates={true} />
              </div>
      )
    }
   }

const mapStateToProps = (state)=>{
     return{
    filters:state.filters
 }
}
const mapDispatchToProps = (dispatch)=>{
   return {
      setEndDate: (endDate)=>dispatch(setEndDate(endDate)),
      setStartDate: (startDate)=>dispatch(setStartDate(startDate)),
      setTextFilter:(text)=>dispatch(setTextFilter(text)),
      sortByDate: ()=>dispatch(sortByDate()),
      sortByAmount:()=>dispatch(sortByAmount())
 
   }
}
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter)