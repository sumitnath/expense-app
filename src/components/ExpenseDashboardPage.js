import React from 'react';
//import { BrowserRouter,Route,Switch,Link,NavLink } from "react-router-dom";
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummery from './ExpenseSummery';

const ExpenseDashboardPage = () => (
  <div>
    This is from my dashboard component
    <ExpenseSummery/>
    <ExpenseListFilter/>
    <ExpenseList/>
  </div>
);


  export default ExpenseDashboardPage;