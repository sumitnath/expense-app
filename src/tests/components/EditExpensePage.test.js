import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

test('should render edit expense ',()=>{
    const editExpense = jest.fn()
    const removeExpense = jest.fn() 
    const history = {push: jest.fn()}
    const wrapper = shallow(<EditExpensePage 
        increment ={editExpense}
        decrement ={removeExpense}
        history={history} 
        expense = {expenses[2]}
        /> )

 expect(wrapper).toMatchSnapshot()
})

test('should handle edit expense with correct information',()=>{
    const editExpense = jest.fn()
    const history = {push: jest.fn()}
    const wrapper = shallow(<EditExpensePage 
        increment ={editExpense}
        history={history} 
        expense = {expenses[2]}
        /> )

     wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id,expenses[2])
})

test('should handle remove expense with correct information',()=>{
    const removeExpense = jest.fn()
    const history = {push: jest.fn()}
    const wrapper = shallow(<EditExpensePage 
        decrement ={removeExpense}
        history={history} 
        expense = {expenses[2]}
        /> )

     wrapper.find('button').prop('onClick')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[2].id})
})