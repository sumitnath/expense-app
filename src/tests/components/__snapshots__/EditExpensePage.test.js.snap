// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense  1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditExpensePage
    decrement={[Function]}
    expense={
        Object {
            "amount": 13,
            "createdAt": 345600000,
            "description": "Credit Card",
            "id": 3,
            "note": "",
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
    increment={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
           Edit Your expense
</h3>,
        <ExpenseForm
          expense={
                    Object {
                              "amount": 13,
                              "createdAt": 345600000,
                              "description": "Credit Card",
                              "id": 3,
                              "note": "",
                            }
          }
          onSubmit={[Function]}
/>,
        <button
          onClick={[Function]}
>
          Remove
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Edit Your expense",
        },
        "ref": null,
        "rendered": " Edit Your expense",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "expense": Object {
            "amount": 13,
            "createdAt": 345600000,
            "description": "Credit Card",
            "id": 3,
            "note": "",
          },
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Remove",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Remove",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
             Edit Your expense
</h3>,
          <ExpenseForm
            expense={
                        Object {
                                    "amount": 13,
                                    "createdAt": 345600000,
                                    "description": "Credit Card",
                                    "id": 3,
                                    "note": "",
                                  }
            }
            onSubmit={[Function]}
/>,
          <button
            onClick={[Function]}
>
            Remove
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Edit Your expense",
          },
          "ref": null,
          "rendered": " Edit Your expense",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "expense": Object {
              "amount": 13,
              "createdAt": 345600000,
              "description": "Credit Card",
              "id": 3,
              "note": "",
            },
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Remove",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Remove",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
